File: arcticnetwork.java
import java.io.*;
import java.util.*;

class DoublePair implements Comparable<DoublePair> {// stores coordinates
    private double _first, _second;
  
    public DoublePair(double f, double s) {
      _first = f;
      _second = s;
    }
  
    public int compareTo(DoublePair o) {
      if (Double.compare(this.first(),o.first()) != 0) return Double.compare(this.first(),o.first());
      else return Double.compare(this.second(),o.second());
    }
  
    public double first() {return _first;}
  
    public double second() {return _second;}
}

class DoubleIntPair implements Comparable<DoubleIntPair> {// stores coordinates
    private double _first;
    private int _second;
  
    public DoubleIntPair(double f, int s) {
      _first = f;
      _second = s;
    }
  
    public int compareTo(DoubleIntPair o) {
      if (Double.compare(this.first(),o.first()) != 0) return Double.compare(this.first(),o.first());
      else return this.second() - o.second();
    }
  
    public double first() {return _first;}
  
    public int second() {return _second;}
}

public class arcticnetwork {
    public static double distance(double x1, double y1, double x2, double y2){
        return Math.sqrt(Math.pow(x1-x2,2) + Math.pow(y1-y2, 2)); //calculate Euclidean distance
    }
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        PrintWriter pw = new PrintWriter(new BufferedWriter(new OutputStreamWriter(System.out)));
        double INF = 1000000000; // use 1 billion to represent infinity
        int n = Integer.parseInt(br.readLine());
        for (int i = 0; i < n; i++){ // use Prim's variant for dense graphs with O(P^2) time complexity
            String[] L = br.readLine().split(" ");
            int S = Integer.parseInt(L[0]), P = Integer.parseInt(L[1]);
            DoublePair[] points = new DoublePair[P];
            DoubleIntPair[] A = new DoubleIntPair[P];
            for (int j = 0; j < P; j++) {
                String[] line = br.readLine().split(" ");
                points[j] = new DoublePair(Double.parseDouble(line[0]), Double.parseDouble(line[1]));
                A[j] = new DoubleIntPair(INF, j); // initialisation
            }
            boolean[] taken = new boolean[P];
            taken[0] = true; A[0] = new DoubleIntPair(0, 0);
            PriorityQueue<Double> cost = new PriorityQueue<Double>(Comparator.reverseOrder()); // turn it into maxHeap; cost keeps track of all edge weights in MST
            int counter = 1;  // counter keeps track of number of vertices in taken set to true
            while (counter < P){ // while not all vertices are in MST
                double lowest = INF; int v = 0; //scan A to get v where A[v].first is minimum in A
                for (int j = 0; j < P; j++){
                    if (A[j].first() < lowest){
                        lowest = A[j].first();
                        v = j;
                    }
                }
                if (!taken[v]) {counter += 1; taken[v] = true;}
                if (!taken[A[v].second()]) {counter += 1; taken[A[v].second()] = true;}
                cost.offer(A[v].first());
                A[v] = new DoubleIntPair(INF, A[v].second()); //prevent picking the same point
                for (int j = 0; j < P; j++){
                    if (!taken[j] && A[j].first() > distance(points[v].first(), points[v].second(), points[j].first(), points[j].second())){
                        A[j] = new DoubleIntPair(distance(points[v].first(), points[v].second(), points[j].first(), points[j].second()), v);
                    }
                }
            }
            for (int j = 0; j < S - 1; j++) {cost.poll();} // with S channels, there can be S - 1 edges. Remove the S-1 edges with the largest weights.
            pw.printf("%.2f%n",cost.peek()); //all remaining edges will have weights < largest edge in PQ
        }
        br.close();
        pw.close();
    }
}

File: arcticnetwork.py
from sys import stdin,stdout
from math import sqrt
from heapq import heappush,heappop,heapify

INF = 10**9 # use one billion to represent infinity
for _ in range(int(stdin.readline())): #Prim's variant for dense graphs is used
    S,P = map(int,stdin.readline().split())
    points = [tuple(map(int,stdin.readline().split())) for _ in range(P)]
    A,taken = [(INF,v) for v in range(P)],[False for _ in range(P)]
    A[0] = (0,0) #start from first point
    taken[0] = True; cost = []; counter = 1
    while counter < P: #O(P)
        lowest,v = INF,0 # scan A to get v where A[v][0] is minimum in A
        for i in range(P): #O(P)
            if A[i][0] < lowest:
                lowest = A[i][0]
                v = i
        #add vertices to taken
        if not taken[v]:
            taken[v] = True
            counter += 1
        if not taken[A[v][1]]:
            taken[A[v][1]] = True
            counter += 1
        cost.append(-A[v][0]) ## since default heap in Python is a min heap, a minus sign is required to turn it into max heap
        A[v] = (INF,A[v][1]) #prevent picking the same point
        for j in range(P): #O(P)
            if not taken[j] and A[j][0] > sqrt((points[v][0]-points[j][0])**2+(points[v][1]-points[j][1])**2):
                A[j] = (sqrt((points[v][0]-points[j][0])**2+(points[v][1]-points[j][1])**2),v)
    heapify(cost) #O(P)
    for i in range(S-1):
        heappop(cost) # with S channels, S - 1 edges can be removed
    stdout.write("{:.2f}".format(-heappop(cost))+'\n') # heappop(cost) to remove largest edge weight D in remaining edges; all remaining edges will have weights < D
    
